code/modules/reagents/chemistry/holder.dm

```diff
@@ -485,44 +558,261 @@
+
+
 						for(var/mob/M in seen)
-							to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
-							ME2.name = "used slime extract"
-							ME2.desc = "This extract has been used up."
+							to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
+
+					if(istype(cached_my_atom, /obj/item/slime_extract))//if there's an extract and it's used up.
+						var/obj/item/slime_extract/ME2 = my_atom
+						ME2.Uses--
+						if(ME2.Uses <= 0) // give the notification that the slime core is dead
+							for(var/mob/M in seen)
+								to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
+								ME2.name = "used slime extract"
+								ME2.desc = "This extract has been used up."
 
-			selected_reaction.on_reaction(src, multiplier)
-			reaction_occurred = 1
+				selected_reaction.on_reaction(src, multiplier, special_react_result)
+				reaction_occurred = 1
+				continue_reacting = TRUE
 
 	while(reaction_occurred)
 	update_total()
 	return 0
 
+/datum/reagents/process()
+	var/datum/chemical_reaction/fermi/C = fermiReactID
+
+	var/list/cached_required_reagents = C.required_reagents//update reagents list
+	var/list/cached_results = C.results//resultant chemical list
+	var/multiplier = INFINITY
+
+	for(var/B in cached_required_reagents) //
+		multiplier = min(multiplier, round((get_reagent_amount(B) / cached_required_reagents[B]), 0.001))
+	if (multiplier == 0)
+		fermiEnd()
+		return
+	for(var/P in cached_results)
+		targetVol = cached_results[P]*multiplier
+
+	if (fermiIsReacting == FALSE)
+		CRASH("Fermi has refused to stop reacting even though we asked her nicely.")
+
+	if (chem_temp > C.OptimalTempMin && fermiIsReacting)//To prevent pointless reactions
+		if( (pH >= (C.OptimalpHMin - C.ReactpHLim)) && (pH <= (C.OptimalpHMax + C.ReactpHLim)) )
+			if (reactedVol < targetVol)
+				reactedVol = fermiReact(fermiReactID, chem_temp, pH, reactedVol, targetVol, cached_required_reagents, cached_results, multiplier)
+			else//Volume is used up
+				fermiEnd()
+				return
+		else//pH is out of range
+			fermiEnd()
+			return
+	else//Temperature is too low, or reaction has stopped.
+		fermiEnd()
+		return
+
+/datum/reagents/proc/fermiEnd()
+	var/datum/chemical_reaction/fermi/C = fermiReactID
+	STOP_PROCESSING(SSprocessing, src)
+	fermiIsReacting = FALSE
+	reactedVol = 0
+	targetVol = 0
+	//pH check, handled at the end to reduce calls.
+	if(istype(my_atom, /obj/item/reagent_containers))
+		var/obj/item/reagent_containers/RC = my_atom
+		RC.pH_check()
+	C.FermiFinish(src, my_atom)
+	handle_reactions()
+	update_total()
+	//Reaction sounds and words
+	playsound(get_turf(my_atom), C.mix_sound, 80, 1)
+	var/list/seen = viewers(5, get_turf(my_atom))
+	var/iconhtml = icon2html(my_atom, seen)
+	for(var/mob/M in seen)
+		to_chat(M, "<span class='notice'>[iconhtml] [C.mix_message]</span>")
```

Comment:
visible message

Reply:
i'm not going to mark any more, consider this a general "convert this into visible message" comment